### ISSUE TITLE
Install failure on Debian Bullseye - missing: SDL2_INCLUDE_DIR

### ISSUE THREAD

**Issue by jrial** [mwydmuch/ViZDoom#440]
### The issue
I was unable to install vizdoom through pip under Python 3. I tried both under a virtualenv using pipenv, and through the Python 3.8.2 that's installed system-wide, but kept getting the same error: `Could NOT find SDL2 (missing: SDL2_INCLUDE_DIR)`

### The solution
After analysing the output from the install failure, I managed to solve it by setting the following environment variable: `export SDL2DIR=/usr/include/x86_64-linux-gnu/SDL2/`

Turns out this path was not among the ones listed in `usr/include/x86_64-linux-gnu/`, perhaps it should be added?

### Version numbers and install logs

System: Debian Bullseye
Python version: 3.8.2
GCC: 9.3.0
Java/Lua: irrelevant

All dependencies installed:
```
[23:59] nighty@obsidian ViZDoom$ echo "Start of output" && for i in g++ make cmake libsdl2-dev git zlib1g-dev libbz2-dev libjpeg-dev libfluidsynth-dev libgme-dev libopenal-dev libmpg123-dev libsndfile1-dev libwildmidi-dev libgtk2.0-dev libgtk-3-dev timidity nasm tar chrpath build-essential unzip libboost-all-dev python3-dev python3-pip ; do dpkg -l |grep "${i}" > /dev/null || echo "${i} is not installed" ; done && echo "End of output"
Start of output
End of output
```

Trying to pip install vizdoom, I got an error about SDL2 not being found:
```
[23:59] nighty@obsidian ViZDoom$ pip3 install vizdoom
Collecting vizdoom
  Using cached vizdoom-1.1.7.tar.gz (4.9 MB)
Requirement already satisfied: numpy in /home/nighty/.local/lib/python3.8/site-packages (from vizdoom) (1.18.2)
Building wheels for collected packages: vizdoom
  Building wheel for vizdoom (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /usr/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-y8ug_xws/vizdoom/setup.py'"'"'; __file__='"'"'/tmp/pip-install-y8ug_xws/vizdoom/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-jhv3v3jl
       cwd: /tmp/pip-install-y8ug_xws/vizdoom/
  Complete output (124 lines):
  running bdist_wheel
  running build
  CMake Warning:
    No source or binary directory provided.  Both will be assumed to be the
    same as the current working directory, but note that this warning will
    become a fatal error in future CMake releases.
  
  
  -- The C compiler identification is GNU 9.3.0
  -- The CXX compiler identification is GNU 9.3.0
  -- Check for working C compiler: /usr/bin/cc
  -- Check for working C compiler: /usr/bin/cc -- works
  -- Detecting C compiler ABI info
  -- Detecting C compiler ABI info - done
  -- Detecting C compile features
  -- Detecting C compile features - done
  -- Check for working CXX compiler: /usr/bin/c++
  -- Check for working CXX compiler: /usr/bin/c++ -- works
  -- Detecting CXX compiler ABI info
  -- Detecting CXX compiler ABI info - done
  -- Detecting CXX compile features
  -- Detecting CXX compile features - done
  -- Looking for pthread.h
  -- Looking for pthread.h - found
  -- Performing Test CMAKE_HAVE_LIBC_PTHREAD
  -- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
  -- Looking for pthread_create in pthreads
  -- Looking for pthread_create in pthreads - not found
  -- Looking for pthread_create in pthread
  -- Looking for pthread_create in pthread - found
  -- Found Threads: TRUE
  -- Found Boost: /usr/include (found version "1.67.0") found components: filesystem thread system date_time chrono regex iostreams atomic
  CMake Deprecation Warning at cmake_modules/CreateLaunchers.cmake:48 (cmake_policy):
    The OLD behavior for policy CMP0026 will be removed from a future version
    of CMake.
  
    The cmake-policies(7) manual explains that the OLD behaviors of all
    policies are deprecated and that a policy should be set to OLD only under
    specific short-term circumstances.  Projects should be ported to the NEW
    behavior and not rely on setting a policy to OLD.
  Call Stack (most recent call first):
    src/vizdoom/CMakeLists.txt:13 (include)
  
  
  -- Found BZip2: /usr/lib/x86_64-linux-gnu/libbz2.so (found version "1.0.8")
  -- Looking for BZ2_bzCompressInit
  -- Looking for BZ2_bzCompressInit - found
  -- Found JPEG: /usr/lib/x86_64-linux-gnu/libjpeg.so (found version "62")
  -- Found ZLIB: /usr/lib/x86_64-linux-gnu/libz.so (found version "1.2.11")
  -- Could NOT find GME (missing: GME_INCLUDE_DIR)
  -- Using system zlib
  -- Using system jpeg library
  -- Using system bzip2 library
  -- Using internal gme library
  -- Performing Test HAVE_NO_ARRAY_BOUNDS
  -- Performing Test HAVE_NO_ARRAY_BOUNDS - Success
  -- Performing Test __LIBGME_TEST_VISIBILITY
  -- Performing Test __LIBGME_TEST_VISIBILITY - Success
  -- Looking for strdup
  -- Looking for strdup - found
  -- Looking for strndup
  -- Looking for strndup - found
  -- Looking for sys/types.h
  -- Looking for sys/types.h - found
  -- Looking for stdint.h
  -- Looking for stdint.h - found
  -- Looking for stddef.h
  -- Looking for stddef.h - found
  -- Check size of char
  -- Check size of char - done
  -- Check size of short
  -- Check size of short - done
  -- Check size of int
  -- Check size of int - done
  -- Check size of long
  -- Check size of long - done
  -- /usr/include
  -- /usr/include
  -- /tmp/pip-install-y8ug_xws/vizdoom/src/vizdoom/lzma/C
  -- Looking for itoa
  -- Looking for itoa - not found
  -- Performing Test DUMB_CAN_USE_SSE
  -- Performing Test DUMB_CAN_USE_SSE - Success
  -- Found PkgConfig: /usr/bin/pkg-config (found version "0.29")
  -- Checking for module 'gtk+-2.0'
  --   Found gtk+-2.0, version 2.24.32
  CMake Error at /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:146 (message):
    Could NOT find SDL2 (missing: SDL2_INCLUDE_DIR)
  Call Stack (most recent call first):
    /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:393 (_FPHSA_FAILURE_MESSAGE)
    cmake_modules/FindSDL2.cmake:179 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
    src/vizdoom/src/CMakeLists.txt:206 (find_package)
  
  
  -- Configuring incomplete, errors occurred!
  See also "/tmp/pip-install-y8ug_xws/vizdoom/CMakeFiles/CMakeOutput.log".
  See also "/tmp/pip-install-y8ug_xws/vizdoom/CMakeFiles/CMakeError.log".
  
  Installation failed, you may be missing some dependencies.
  Please check https://github.com/mwydmuch/ViZDoom/blob/master/doc/Building.md for details
  
  Traceback (most recent call last):
    File "<string>", line 1, in <module>
    File "/tmp/pip-install-y8ug_xws/vizdoom/setup.py", line 91, in <module>
      setup(
    File "/usr/lib/python3/dist-packages/setuptools/__init__.py", line 145, in setup
      return distutils.core.setup(**attrs)
    File "/usr/lib/python3.8/distutils/core.py", line 148, in setup
      dist.run_commands()
    File "/usr/lib/python3.8/distutils/dist.py", line 966, in run_commands
      self.run_command(cmd)
    File "/usr/lib/python3.8/distutils/dist.py", line 985, in run_command
      cmd_obj.run()
    File "/usr/lib/python3/dist-packages/wheel/bdist_wheel.py", line 223, in run
      self.run_command('build')
    File "/usr/lib/python3.8/distutils/cmd.py", line 313, in run_command
      self.distribution.run_command(command)
    File "/usr/lib/python3.8/distutils/dist.py", line 985, in run_command
      cmd_obj.run()
    File "/tmp/pip-install-y8ug_xws/vizdoom/setup.py", line 81, in run
      subprocess.check_call(['cmake'] + cmake_arg_list)
    File "/usr/lib/python3.8/subprocess.py", line 364, in check_call
      raise CalledProcessError(retcode, cmd)
  subprocess.CalledProcessError: Command '['cmake', '-DCMAKE_BUILD_TYPE=Release', '-DBUILD_PYTHON=ON', '-DPYTHON_EXECUTABLE=/usr/bin/python3.8', '-DBUILD_PYTHON3=ON']' returned non-zero exit status 1.
  ----------------------------------------
  ERROR: Failed building wheel for vizdoom
  Running setup.py clean for vizdoom
Failed to build vizdoom
Installing collected packages: vizdoom
    Running setup.py install for vizdoom ... error
    ERROR: Command errored out with exit status 1:
     command: /usr/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-y8ug_xws/vizdoom/setup.py'"'"'; __file__='"'"'/tmp/pip-install-y8ug_xws/vizdoom/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-nadwoa9x/install-record.txt --single-version-externally-managed --user --prefix= --compile --install-headers /home/nighty/.local/include/python3.8/vizdoom
         cwd: /tmp/pip-install-y8ug_xws/vizdoom/
    Complete output (126 lines):
    running install
    running build
    CMake Warning:
      No source or binary directory provided.  Both will be assumed to be the
      same as the current working directory, but note that this warning will
      become a fatal error in future CMake releases.
    
    
    -- The C compiler identification is GNU 9.3.0
    -- The CXX compiler identification is GNU 9.3.0
    -- Check for working C compiler: /usr/bin/cc
    -- Check for working C compiler: /usr/bin/cc -- works
    -- Detecting C compiler ABI info
    -- Detecting C compiler ABI info - done
    -- Detecting C compile features
    -- Detecting C compile features - done
    -- Check for working CXX compiler: /usr/bin/c++
    -- Check for working CXX compiler: /usr/bin/c++ -- works
    -- Detecting CXX compiler ABI info
    -- Detecting CXX compiler ABI info - done
    -- Detecting CXX compile features
    -- Detecting CXX compile features - done
    -- Looking for pthread.h
    -- Looking for pthread.h - found
    -- Performing Test CMAKE_HAVE_LIBC_PTHREAD
    -- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
    -- Looking for pthread_create in pthreads
    -- Looking for pthread_create in pthreads - not found
    -- Looking for pthread_create in pthread
    -- Looking for pthread_create in pthread - found
    -- Found Threads: TRUE
    -- Found Boost: /usr/include (found version "1.67.0") found components: filesystem thread system date_time chrono regex iostreams atomic
    CMake Deprecation Warning at cmake_modules/CreateLaunchers.cmake:48 (cmake_policy):
      The OLD behavior for policy CMP0026 will be removed from a future version
      of CMake.
    
      The cmake-policies(7) manual explains that the OLD behaviors of all
      policies are deprecated and that a policy should be set to OLD only under
      specific short-term circumstances.  Projects should be ported to the NEW
      behavior and not rely on setting a policy to OLD.
    Call Stack (most recent call first):
      src/vizdoom/CMakeLists.txt:13 (include)
    
    
    -- Found BZip2: /usr/lib/x86_64-linux-gnu/libbz2.so (found version "1.0.8")
    -- Looking for BZ2_bzCompressInit
    -- Looking for BZ2_bzCompressInit - found
    -- Found JPEG: /usr/lib/x86_64-linux-gnu/libjpeg.so (found version "62")
    -- Found ZLIB: /usr/lib/x86_64-linux-gnu/libz.so (found version "1.2.11")
    -- Could NOT find GME (missing: GME_INCLUDE_DIR)
    -- Using system zlib
    -- Using system jpeg library
    -- Using system bzip2 library
    -- Using internal gme library
    -- Performing Test HAVE_NO_ARRAY_BOUNDS
    -- Performing Test HAVE_NO_ARRAY_BOUNDS - Success
    -- Performing Test __LIBGME_TEST_VISIBILITY
    -- Performing Test __LIBGME_TEST_VISIBILITY - Success
    -- Looking for strdup
    -- Looking for strdup - found
    -- Looking for strndup
    -- Looking for strndup - found
    -- Looking for sys/types.h
    -- Looking for sys/types.h - found
    -- Looking for stdint.h
    -- Looking for stdint.h - found
    -- Looking for stddef.h
    -- Looking for stddef.h - found
    -- Check size of char
    -- Check size of char - done
    -- Check size of short
    -- Check size of short - done
    -- Check size of int
    -- Check size of int - done
    -- Check size of long
    -- Check size of long - done
    -- /usr/include
    -- /usr/include
    -- /tmp/pip-install-y8ug_xws/vizdoom/src/vizdoom/lzma/C
    -- Looking for itoa
    -- Looking for itoa - not found
    -- Performing Test DUMB_CAN_USE_SSE
    -- Performing Test DUMB_CAN_USE_SSE - Success
    -- Found PkgConfig: /usr/bin/pkg-config (found version "0.29")
    -- Checking for module 'gtk+-2.0'
    --   Found gtk+-2.0, version 2.24.32
    CMake Error at /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:146 (message):
      Could NOT find SDL2 (missing: SDL2_INCLUDE_DIR)
    Call Stack (most recent call first):
      /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:393 (_FPHSA_FAILURE_MESSAGE)
      cmake_modules/FindSDL2.cmake:179 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
      src/vizdoom/src/CMakeLists.txt:206 (find_package)
    
    
    -- Configuring incomplete, errors occurred!
    See also "/tmp/pip-install-y8ug_xws/vizdoom/CMakeFiles/CMakeOutput.log".
    See also "/tmp/pip-install-y8ug_xws/vizdoom/CMakeFiles/CMakeError.log".
    
    Installation failed, you may be missing some dependencies.
    Please check https://github.com/mwydmuch/ViZDoom/blob/master/doc/Building.md for details
    
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-install-y8ug_xws/vizdoom/setup.py", line 91, in <module>
        setup(
      File "/usr/lib/python3/dist-packages/setuptools/__init__.py", line 145, in setup
        return distutils.core.setup(**attrs)
      File "/usr/lib/python3.8/distutils/core.py", line 148, in setup
        dist.run_commands()
      File "/usr/lib/python3.8/distutils/dist.py", line 966, in run_commands
        self.run_command(cmd)
      File "/usr/lib/python3.8/distutils/dist.py", line 985, in run_command
        cmd_obj.run()
      File "/usr/lib/python3/dist-packages/setuptools/command/install.py", line 61, in run
        return orig.install.run(self)
      File "/usr/lib/python3.8/distutils/command/install.py", line 589, in run
        self.run_command('build')
      File "/usr/lib/python3.8/distutils/cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "/usr/lib/python3.8/distutils/dist.py", line 985, in run_command
        cmd_obj.run()
      File "/tmp/pip-install-y8ug_xws/vizdoom/setup.py", line 81, in run
        subprocess.check_call(['cmake'] + cmake_arg_list)
      File "/usr/lib/python3.8/subprocess.py", line 364, in check_call
        raise CalledProcessError(retcode, cmd)
    subprocess.CalledProcessError: Command '['cmake', '-DCMAKE_BUILD_TYPE=Release', '-DBUILD_PYTHON=ON', '-DPYTHON_EXECUTABLE=/usr/bin/python3.8', '-DBUILD_PYTHON3=ON']' returned non-zero exit status 1.
    ----------------------------------------
ERROR: Command errored out with exit status 1: /usr/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-y8ug_xws/vizdoom/setup.py'"'"'; __file__='"'"'/tmp/pip-install-y8ug_xws/vizdoom/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-nadwoa9x/install-record.txt --single-version-externally-managed --user --prefix= --compile --install-headers /home/nighty/.local/include/python3.8/vizdoom Check the logs for full command output.
```

**Comment by Miffyli** [comment]
Hmm SDL2 includes are under `/usr/include/SDL2` for me (Ubuntu 18.04), but I also have the `x86_64-linux-gnu` directory. I do not know why it is under different directory for you. Is this perhaps something that will happen with future Debian-based distributions? In that case a check for this path would be useful, so people do not have to manually export things.

**Comment by jrial** [comment]
Under `/usr/include/SDL2` I have some header files: `SDL_mixer.h` and `SDL_ttf.h`. These belong to the packages `libsdl2-mixer-dev:amd64` and `libsdl2-ttf-dev:amd64` respectively.

From what I gather from `cmake_modules/FindSDL2.cmake` however, we're looking for `SDL.h`, which is in `/usr/include/x86_64-linux-gnu/SDL2/` and belongs to the package `libsdl2-dev:amd64`. 

* In Debian Buster (stable), this file lives under `/usr/include/SDL2`; see [here](https://packages.debian.org/buster/amd64/libsdl2-dev/filelist).
* In Bullseye (testing) it got moved under `/usr/include/x86_64-linux-gnu/SDL2`, as can be verified [here](https://packages.debian.org/bullseye/amd64/libsdl2-dev/filelist).
* For i386, the includes for Buster also lived in `usr/include/SDL2/`, as you can see [here](https://packages.debian.org/buster/i386/libsdl2-dev/filelist).
* In Bullseye they also moved, this time to `/usr/include/i386-linux-gnu/SDL2/`, as can be verified [here](https://packages.debian.org/bullseye/i386/libsdl2-dev/filelist).

It appears Debian is going to put includes in an architecture-specific include folder with the next release. Not sure which architectures of theirs ViZDoom supports, but you can check where the include files for every architecture are located [at the bottom of this page](https://packages.debian.org/bullseye/libsdl2-dev).

From what I gather online though, it appears that the proper way to determine where the SDL2 includes live is by using the output of `sdl2-config --cflags`, which on my system results in: `-D_REENTRANT -I/usr/include/x86_64-linux-gnu/SDL2 -I/usr/include/SDL2`.

**Comment by Miffyli** [comment]
Sorry for the slow reply.

@mwydmuch Can you take a look at this? You are the one more familiar with all of this ^^

**Comment by mwydmuch** [comment]
Hi @jrial thank you for reporting the problem and also sorry for the late response. It seems to me that we should add these new paths to `cmake_modules/FindSDL2.cmake` to support Bullseye.

**Comment by mwydmuch** [comment]
Hi @jrial, I wanted to fix it but I visited the link provided by you: https://packages.debian.org/bullseye/amd64/libsdl2-dev/filelist and I don't see how the directories changed, to me, it looks like SDL.h is still under /usr/include/SDL2. Am I missing something, or maybe something changed about SDL2 package?

**Comment by jrial** [comment]
Seems like it got changed back.

From the changelog:

> libsdl2 (2.0.10+dfsg1-3) unstable; urgency=medium
>  * Team upload
>
>  [ Gianfranco Costamagna ]
>  * Fix autopkgtests for cross-architecture testing, by making sure
>    sdl2-config calls the correct pkg-config command
>
>  [ Simon McVittie ]
>  * Revert "Add the non-multiarch include dir to pkg-config cflags for
>    compatibility."
>  * Revert "Install all headers in an architecture-specific location"
>  * Make SDL_config.h redirect to an architecture-specific version
>    (Closes: #909740, #951087, #951943, #951959, #951974, #951976, #952049,
>    #951962, #952105, #955416)
>  * Create a SDL_platform.h symlink in /usr/include/MULTIARCH/SDL2
>    (Closes: #952046, #952066, #952098)
>  * Add autopkgtests for various deprecated use patterns

And indeed, after an upgrade, the header file is back to its original location:

```
[09:10] nighty@obsidian ~$ find /usr/include -name SDL.h
/usr/include/x86_64-linux-gnu/SDL2/SDL.h
[09:10] nighty@obsidian ~$ dpkg -S /usr/include/x86_64-linux-gnu/SDL2/SDL.h
libsdl2-dev:amd64: /usr/include/x86_64-linux-gnu/SDL2/SDL.h
[09:10] nighty@obsidian ~$ sudo apt-get install --only-upgrade libsdl2-dev
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following packages were automatically installed and are no longer required:
  gir1.2-mutter-5 libboost-date-time1.71.0 libboost-filesystem1.71.0 libboost-iostreams1.71.0 libboost-locale1.71.0 libboost-thread1.71.0 libdvdread7 libmozjs-60-0 libmutter-5-0 libpipewire-0.2-1 libpython3.7 libpython3.7-minimal libpython3.7-stdlib libqpdf26 libruby2.5
  ruby2.5
Use 'sudo apt autoremove' to remove them.
The following additional packages will be installed:
  libsdl2-2.0-0
The following packages will be upgraded:
  libsdl2-2.0-0 libsdl2-dev
2 upgraded, 0 newly installed, 0 to remove and 135 not upgraded.
Need to get 0 B/1,153 kB of archives.
After this operation, 9,216 B of additional disk space will be used.
Do you want to continue? [Y/n] 
Reading changelogs... Done
(Reading database ... 284230 files and directories currently installed.)
Preparing to unpack .../libsdl2-dev_2.0.10+dfsg1-3_amd64.deb ...
Unpacking libsdl2-dev:amd64 (2.0.10+dfsg1-3) over (2.0.10+dfsg1-2) ...
Preparing to unpack .../libsdl2-2.0-0_2.0.10+dfsg1-3_amd64.deb ...
Unpacking libsdl2-2.0-0:amd64 (2.0.10+dfsg1-3) over (2.0.10+dfsg1-2) ...
Setting up libsdl2-2.0-0:amd64 (2.0.10+dfsg1-3) ...
Setting up libsdl2-dev:amd64 (2.0.10+dfsg1-3) ...
Processing triggers for man-db (2.9.1-1) ...
Processing triggers for libc-bin (2.30-4) ...
[09:10] nighty@obsidian ~$ find /usr/include -name SDL.h
/usr/include/SDL2/SDL.h
```

I'm closing this.

**Comment by mwydmuch** [comment]
Good, thank you @jrial for confirming!
